### PR TITLE
Dart Sass Support

### PR DESCRIPTION
**Pull request by kaelig**
Support the new Dart Sass implementation – https://github.com/sass/dart-sass

Tip to run tests every time an scss file is modified:

```
npm i -g nodemon
nodemon -e scss -x "/path/to/dart-sass test/test.scss --trace" test
```

### CODE REVIEW

**Comment by kaelig** [comment]
FYI @nex3: this is what I get when trying to compile tests in Dart Sass:

https://travis-ci.org/sass-mq/sass-mq/jobs/172138466#L593

```log
Unexpected exception:RangeError (index): Index out of range: index should be less than 1: 1
node_modules/dart-sass/sass.dart.js 4154:17   dart.wrapException
node_modules/dart-sass/sass.dart.js 4104:15   dart.ioore
node_modules/dart-sass/sass.dart.js 17629:22  Environment.dart.Environment.getVariable$1
node_modules/dart-sass/sass.dart.js 27366:37  _PerformVisitor.dart._PerformVisitor.visitVariableDeclaration$1
node_modules/dart-sass/sass.dart.js 16637:24  VariableDeclaration.dart.VariableDeclaration.accept$1
node_modules/dart-sass/sass.dart.js 31349:41  J.accept$1$x
node_modules/dart-sass/sass.dart.js 28126:13  _PerformVisitor__visitDynamicImport__closure.dart._PerformVisitor__visitDynamicImport__closure.call$0
node_modules/dart-sass/sass.dart.js 27703:27  _PerformVisitor.dart._PerformVisitor._withEnvironment$2
node_modules/dart-sass/sass.dart.js 28118:12  _PerformVisitor__visitDynamicImport_closure.dart._PerformVisitor__visitDynamicImport_closure.call$0
node_modules/dart-sass/sass.dart.js 27207:77  _PerformVisitor.dart._PerformVisitor._visitDynamicImport$1
```

The test is simply `dart-sass test/test.scss` (for now).

Is there a workaround/am I doing anything wrong?

**Comment by nex3** [comment]
That's probably a bug. Can you try running with the pure Dart version and seeing what stack trace you get?

**Comment by nex3** [comment]
FYI I also filed sass/dart-sass#3 to track better stack traces for the JS version.

**Comment by kaelig** [comment]
Thanks! Here is the trace with the pure Dart version:

```log
/Users/kdeloumeauprigent/Downloads/dart-sass/dart-sass  test.scss
Unexpected exception:
RangeError (index): Invalid value: Only valid value is 0: 1

dart:core                                                         List.[]
package:sass/src/environment.dart 138                             Environment.getVariable
package:sass/src/visitor/perform.dart 778                         _PerformVisitor.visitVariableDeclaration
package:sass/src/ast/sass/statement/variable_declaration.dart 39  VariableDeclaration.accept
package:sass/src/visitor/perform.dart 504                         _PerformVisitor._visitDynamicImport.<fn>.<fn>
package:sass/src/visitor/perform.dart 1275                        _PerformVisitor._withEnvironment
package:sass/src/visitor/perform.dart 502                         _PerformVisitor._visitDynamicImport.<fn>
package:sass/src/visitor/perform.dart 1360                        _PerformVisitor._withStackFrame
package:sass/src/visitor/perform.dart 501                         _PerformVisitor._visitDynamicImport
package:sass/src/visitor/perform.dart 490                         _PerformVisitor.visitImportRule
package:sass/src/ast/sass/statement/import_rule.dart 22           ImportRule.accept
package:sass/src/visitor/perform.dart 157                         _PerformVisitor.visitStylesheet
package:sass/src/visitor/perform.dart 147                         _PerformVisitor.run
package:sass/src/visitor/perform.dart 46                          evaluate
package:sass/sass.dart 24                                         render
package:sass/src/executable.dart 50                               main
```

**Comment by nex3** [comment]
Hmm, looking at the code, it's not obvious to me how that would happen. If you can find a small reproduction, let me know, otherwise hopefully something in sass-spec will hit it too.

File: _mq.scss [hunk 1]
```diff
@@ -282,6 +282,6 @@ $mq-media-type: all !default;
     }
 }
 
-@if length($mq-show-breakpoints) > 0 {
+@if $mq-show-breakpoints != () {
```
Comment (kaelig) [inline _mq.scss]: @nex3 Somehow this fixes the first error (`Unexpected exception:RangeError (index): Index out of range: index should be less than 1: 1`).

**Comment by kaelig** [comment]
@nex3 I managed to solve the first issue when running `dart-sass test/test.scss --trace`, but then I got this:

```
Unexpected exception:
RangeError (index): Invalid value: Only valid value is 0: 1

dart:core                                                         List.[]
package:sass/src/environment.dart 138                             Environment.getVariable
package:sass/src/visitor/perform.dart 778                         _PerformVisitor.visitVariableDeclaration
package:sass/src/ast/sass/statement/variable_declaration.dart 39  VariableDeclaration.accept
package:sass/src/visitor/perform.dart 504                         _PerformVisitor._visitDynamicImport.<fn>.<fn>
package:sass/src/visitor/perform.dart 1275                        _PerformVisitor._withEnvironment
package:sass/src/visitor/perform.dart 502                         _PerformVisitor._visitDynamicImport.<fn>
package:sass/src/visitor/perform.dart 1360                        _PerformVisitor._withStackFrame
package:sass/src/visitor/perform.dart 501                         _PerformVisitor._visitDynamicImport
package:sass/src/visitor/perform.dart 490                         _PerformVisitor.visitImportRule
package:sass/src/ast/sass/statement/import_rule.dart 22           ImportRule.accept
package:sass/src/visitor/perform.dart 157                         _PerformVisitor.visitStylesheet
package:sass/src/visitor/perform.dart 147                         _PerformVisitor.run
package:sass/src/visitor/perform.dart 46                          evaluate
package:sass/sass.dart 24                                         render
package:sass/src/executable.dart 50                               main
```

Following this, I commented a few lines to force the tests to pass, see 2888930.

Sorry if this is a lot of code to digest. I'm a bit confused because every time I try to make a reduced test case I can't reproduce the bug!

Let me know how I can help further.

**Comment by nex3** [comment]
I *think* this should be fixed with 1.0.0-alpha.2, which I just released.

**Comment by kaelig** [comment]
@nex3 it works indeed, thanks for the quick turnaround!

The output is now exactly the same as Ruby Sass (expect for colors which are lowercase hex in Dart Sass).

**Pull request merged**